semtypes/src/main/java/io/ballerina/types/Core.java

```diff
@@ -89,45 +86,45 @@ public static SubtypeData getComplexSubtypeData(ComplexSemType t, UniformTypeCod
     }
 
     public static SemType union(SemType t1, SemType t2) {
-        UniformTypeBitSet all1;
-        UniformTypeBitSet all2;
-        UniformTypeBitSet some1;
-        UniformTypeBitSet some2;
-
-        if (t1 instanceof UniformTypeBitSet) {
-            if (t2 instanceof UniformTypeBitSet) {
-                return UniformTypeBitSet.from(((UniformTypeBitSet) t1).bitset | ((UniformTypeBitSet) t2).bitset);
+        BasicTypeBitSet all1;
+        BasicTypeBitSet all2;
+        BasicTypeBitSet some1;
+        BasicTypeBitSet some2;
+
+        if (t1 instanceof BasicTypeBitSet) {
+            if (t2 instanceof BasicTypeBitSet) {
+                return BasicTypeBitSet.from(((BasicTypeBitSet) t1).bitset | ((BasicTypeBitSet) t2).bitset);
```

Comment:
Since we refactoring these parts anyway shall we also consider about using pattern matching instead of casting (should also help us avoid casting in the below lines as well)
```suggestion
        if (t1 instanceof BasicTypeBitSet b1) {
            if (t2 instanceof BasicTypeBitSet b2) {
                return BasicTypeBitSet.from(b1.bitset | (b2.bitset);
```